File: 023_Find the Index of the First Occurrence in a String.py
"""
Find the Index of the First Occurrence in a String

Return index of first occurrence of needle in haystack.
"""

def str_str(haystack: str, needle: str) -> int:
    if needle == "":
        return 0
    return haystack.find(needle)


if __name__ == '__main__':
    print('StrStr 1:', str_str('hello','ll'))
    print('StrStr 2:', str_str('aaaaa','bba'))
    print('StrStr 3:', str_str('',''))

# Time Complexity: O(n*m) worst-case, KMP can do O(n+m)
# Space Complexity: O(1)

File: 023_Find the Index of the First Occurrence in a String.cpp
/*
Find the Index of the First Occurrence in a String

Return the index of the first occurrence of needle in haystack, or -1 if not found.

Strategy: Use std::string::find (or implement KMP for linear worst-case). We'll use find for simplicity.
*/

#include <string>
#include <iostream>
using namespace std;

int strStr(const string& haystack, const string& needle) {
    if (needle.empty()) return 0;
    size_t pos = haystack.find(needle);
    if (pos == string::npos) return -1;
    return (int)pos;
}

int main() {
    cout << "StrStr 1: " << strStr("hello","ll") << '\n';
    cout << "StrStr 2: " << strStr("aaaaa","bba") << '\n';
    cout << "StrStr 3: " << strStr("","") << '\n';
    return 0;
}

/*
Time Complexity: O(n*m) in worst-case for naive, but std::string::find is optimized; KMP gives O(n+m)
Space Complexity: O(1)
*/
